crates/red_knot_python_semantic/src/import_resolution.rs

```diff
@@ -0,0 +1,104 @@
+//! Utilities for resolving import statements that are used from the `types` submodule and the `symbol` submodule.
+
+use ruff_db::files::File;
+use ruff_python_ast as ast;
+
+use crate::{
+    db::Db,
+    module_name::{ModuleName, ModuleNameResolutionError},
+    module_resolver::resolve_module,
+    semantic_index::{definition::StarImportDefinitionKind, symbol::SymbolTable},
+    symbol::SymbolAndQualifiers,
+    types::Type,
+};
+
+pub(crate) fn resolve_star_import_definition<'db>(
+    db: &'db dyn Db,
+    file: File,
+    star_import_definition: &StarImportDefinitionKind,
+    symbol_table: &SymbolTable,
+) -> Result<SymbolAndQualifiers<'db>, UnresolvedImportFromError> {
+    let import_from = star_import_definition.import();
+    let alias = star_import_definition.alias();
+    let symbol_id = star_import_definition.symbol_id();
+
+    let (_, module_type) = resolve_import_from_module(db, file, import_from, alias)?;
+    let defined_name = symbol_table.symbol(symbol_id).name();
+    let imported_symbol = module_type.member(db, defined_name);
+
+    Ok(imported_symbol)
+}
+
+/// Resolve the [`ModuleName`], and the type of the module, being referred to by an
+/// [`ast::StmtImportFrom`] node. Emit a diagnostic if the module cannot be resolved.
+pub(crate) fn resolve_import_from_module<'db>(
+    db: &'db dyn Db,
+    file: File,
+    import_from: &ast::StmtImportFrom,
+    alias: &ast::Alias,
+) -> Result<(ModuleName, Type<'db>), UnresolvedImportFromError> {
+    let ast::StmtImportFrom { module, level, .. } = import_from;
+    let module = module.as_deref();
+
+    tracing::trace!(
+        "Resolving imported object `{}` from module `{}` into file `{}`",
+        alias.name,
+        format_import_from_module(*level, module),
+        file.path(db),
+    );
+
+    let module_name =
+        ModuleName::from_import_statement(db, file, import_from).map_err(|err| match err {
+            ModuleNameResolutionError::InvalidSyntax => {
+                tracing::debug!("Failed to resolve import due to invalid syntax");
+                UnresolvedImportFromError::InvalidSyntax
+            }
+            ModuleNameResolutionError::TooManyDots => {
+                tracing::debug!(
+                    "Relative module resolution `{}` failed: too many leading dots",
+                    format_import_from_module(*level, module),
+                );
+                UnresolvedImportFromError::UnresolvedModule
+            }
+            ModuleNameResolutionError::UnknownCurrentModule => {
+                tracing::debug!(
+                "Relative module resolution `{}` failed; could not resolve file `{}` to a module",
+                format_import_from_module(*level, module),
+                file.path(db)
+            );
+                UnresolvedImportFromError::UnresolvedModule
+            }
+        })?;
```

Comment:
not sure if we want the tracing every time the function is called or only when this function is called from `infer.rs`? I'm guessing probably the latter?

Reply:
Debug is also user facing (if that helps making a decision)

Reply:
It would be nice if the tracing calls would be implicit from the API (where infer uses one API and the other call site the other). E.g. could `UnresolvedImportError` wrap `ModuleNameResolutionError`? Or we return a type that allows converting from the error returned here to `UnresolvedImportError` where it has two methods. 

Or just add a boolean flag 😆